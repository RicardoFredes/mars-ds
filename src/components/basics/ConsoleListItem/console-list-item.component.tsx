import type { ConsoleListItemProps } from "./console-list-item.types";

import classNames from "classnames";

import Icon from "@/components/basics/Icon";
import Caption from "@/components/typographies/Caption";
import Text from "@/components/typographies/Text";

const ConsoleListItem = ({
  title,
  subtitle,
  iconName: icon,
  isActive,
  isDisabled,
  isConnectorDisabled,
  isAnswered,
  href,
  LinkComponent,
  className,
  ...props
}: ConsoleListItemProps) => {
  const Component = LinkComponent ? LinkComponent : href ? "a" : "div";

  const cn = classNames(
    "console-list-item",
    {
      "console-list-item--is-active": isActive,
      "console-list-item--is-disabled": isDisabled,
      "console-list-item--is-connector-disabled": isConnectorDisabled,
      "console-list-item--is-answered": isAnswered,
    },
    className
  );

  return (
    <Component data-testid="console-list-item" className={cn} href={href} {...props}>
      <div className="console-list-item__icon-container">
        <Icon name={icon} />
      </div>

      <div className="console-list-item__content">
        <Text as="h2">{title}</Text>
        {subtitle && <Caption className="console-list-item__subtitle">{subtitle}</Caption>}
      </div>
    </Component>
  );
};

export default ConsoleListItem;
