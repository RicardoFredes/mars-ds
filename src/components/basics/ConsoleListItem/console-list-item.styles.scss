.console-list-item {
  display: flex;
  align-items: center;
  width: 100%;
  height: 72px;
  color: var(--dark-color-text-primary) !important;
  background-color: var(--dark-color-background-primary);
  transition: background-color ease 0.2s;

  a {
    text-decoration: none;
  }

  &:not(&--is-disabled):hover {
    text-decoration: none;
    cursor: pointer;
    background-color: var(--dark-color-background-secondary);
  }

  &__icon-container {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 32px;
    height: 32px;
    margin: 0 $spacing-lg;
    background-color: $color-neutral-900;
    border-radius: $border-radius-lg;
    transition: all ease 0.2s;

    &::before,
    &::after {
      position: absolute;
      left: 50%;
      width: 1px;
      height: 12px;
      margin-left: -1px;
      content: "";
      background-color: $color-divider;
    }

    &::before {
      bottom: 100%;
      margin-bottom: $spacing-sm;
    }

    &::after {
      top: 100%;
      margin-top: $spacing-sm;
    }
  }

  &__content {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
  }

  &__subtitle {
    color: var(--dark-color-text-secondary);
  }

  &--is-active {
    cursor: pointer;
    background-color: var(--dark-color-background-secondary);

    .console-list-item__icon-container {
      outline: 2px solid var(--dark-color-text-primary);
    }
  }

  &--is-disabled {
    text-decoration: none !important;
    cursor: default;
    background-color: var(--dark-color-background-primary);

    .console-list-item__icon-container {
      outline: none;
    }
  }

  &--is-connector-disabled {
    cursor: default;

    .console-list-item__icon-container {
      outline: none;

      &::before,
      &::after {
        display: none;
      }
    }
  }

  &--is-answered {
    cursor: default;
    background-color: var(--dark-color-background-primary);

    .console-list-item__icon-container {
      color: $color-neutral-900;
      background-color: $color-white;
      outline: none;
    }
  }
}

.theme-dark {
  .console-list-item {
    &__icon-container {
      background-color: $color-neutral-100;
    }

    &--is-answered {
      .console-list-item__icon-container {
        color: $color-neutral-100;
        background-color: $color-white;
        outline: none;
      }
    }
  }
}
