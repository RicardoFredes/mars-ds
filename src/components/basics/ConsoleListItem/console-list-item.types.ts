import type { AnchorHTMLAttributes, ElementType } from "react";

interface ConsoleListItemProps
  extends Pick<AnchorHTMLAttributes<HTMLAnchorElement>, "href" | "target"> {
  title: string;
  subtitle?: string;
  iconName: string;
  isActive?: boolean;
  isDisabled?: boolean;
  isConnectorDisabled?: boolean;
  isAnswered?: boolean;
  className?: string;
  LinkComponent?: ElementType | string;
}

export type { ConsoleListItemProps };
