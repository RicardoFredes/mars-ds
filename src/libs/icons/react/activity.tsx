/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconActivity = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12" />
</svg>
);
  
export default IconActivity;