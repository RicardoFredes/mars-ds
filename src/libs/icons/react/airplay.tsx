/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAirplay = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  stroke-linejoin="round"
>
  <path d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1" />
  <polygon points="12 15 17 21 7 21 12 15" />
</svg>
);
  
export default IconAirplay;