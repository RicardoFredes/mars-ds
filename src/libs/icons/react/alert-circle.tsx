/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAlertCircle = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <circle cx="12" cy="12" r="10" />
  <line x1="12" y1="8" x2="12" y2="12" />
  <line x1="12" y1="16" x2="12.01" y2="16" />
</svg>
);
  
export default IconAlertCircle;