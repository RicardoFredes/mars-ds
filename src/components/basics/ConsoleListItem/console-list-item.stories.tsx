import type { ConsoleListItemProps } from "./console-list-item.types";
import type { ComponentMeta, ComponentStory } from "@storybook/react";

import ConsoleListItem from "./console-list-item.component";

const myStoryFunctionDescription =
  "The ConsoleListItem component is meant to be used by the ConsoleList component, either alone or inside a ConsoleGroupProps component.";

export default {
  title: "Components/Console/ConsoleListItem",
  parameters: {
    docs: {
      storyDescription: myStoryFunctionDescription,
    },
  },
  component: ConsoleListItem,
  argTypes: {
    title: {
      description: "The title to be shown inside the component",
    },
    subtitle: {
      description: "The subtitle to be shown below the title",
    },
    iconName: {
      description: "The icon that is shown next to the title/subtitle",
    },
    isActive: {
      description:
        "Whether the item is active (has the same behavior of when the item is hovered) or not",
      control: "boolean",
      defaultValue: false,
    },
    isConnectorDisabled: {
      description:
        "Whether the vertical lines connecting the component to the component above and below is disabled or not",
      control: "boolean",
      defaultValue: false,
    },
    isAnswered: {
      description:
        "Whether the item is checked or not, this causes it to show a different icon, with a different color",
      control: "boolean",
      defaultValue: false,
    },
    isFirstItem: {
      description:
        "Whether or not to show the vertical line connecting the component to the component above",
      control: "boolean",
      defaultValue: false,
    },
    isLastItem: {
      description:
        "Whether or not to show the vertical line connecting the component to the component below",
      control: "boolean",
      defaultValue: false,
    },
    href: {
      description:
        "Transforms the underlying component into an anchor tag, redirecting to the provided URL",
      control: {
        type: "select",
        /** @import("react").HTMLAttributeAnchorTarget */
        options: ["", "_self", "_blank", "_parent", "_top"],
      },
    },
    as: {
      description: "Transforms the underlying component into a custom component",
    },
  },
} as ComponentMeta<typeof ConsoleListItem>;

const Template: ComponentStory<typeof ConsoleListItem> = (args) => <ConsoleListItem {...args} />;

const DefaultArgs: ConsoleListItemProps = {
  title: "Item de console",
  subtitle: "99 quest√µes",
  iconName: "edit",
};

export const Default = Template.bind({});
Default.args = DefaultArgs;
Default.parameters = {
  docs: {
    description: {
      story: "Component with the default props (all modifiers are disabled)",
    },
  },
};

export const Active = Template.bind({});
Active.args = { ...DefaultArgs, isActive: true };
Active.parameters = {
  docs: {
    description: {
      story: "Component with the isActive modifier enabled",
    },
  },
};

export const Disabled = Template.bind({});
Disabled.args = { ...DefaultArgs, isDisabled: true };
Disabled.parameters = {
  docs: {
    description: {
      story: "Component with the isDisabled modifier enabled",
    },
  },
};

export const ConnectorDisabled = Template.bind({});
ConnectorDisabled.args = { ...DefaultArgs, isConnectorDisabled: true };
ConnectorDisabled.parameters = {
  docs: {
    description: {
      story: "Component with the isConnectorDisabled modifier enabled",
    },
  },
};

export const Answered = Template.bind({});
Answered.args = { ...DefaultArgs, isAnswered: true };
Answered.parameters = {
  docs: {
    description: {
      story: "Component with the isAnswered modifier enabled",
    },
  },
};

export const WithLink = Template.bind({});
WithLink.args = { ...DefaultArgs, href: "https://www.mesalva.com", target: "_blank" };
WithLink.parameters = {
  docs: {
    description: {
      story: "Component as a link",
    },
  },
};
