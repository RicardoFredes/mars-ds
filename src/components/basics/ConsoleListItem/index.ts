/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 * run "yarn reindex" to update
 */

import ConsoleListItem from "./console-list-item.component";
export default ConsoleListItem;
export * from "./console-list-item.types";
