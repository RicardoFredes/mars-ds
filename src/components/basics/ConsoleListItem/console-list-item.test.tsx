import type { ConsoleListItemProps } from "./console-list-item.types";

import { render } from "@testing-library/react";

import ConsoleListItem from "./console-list-item.component";

const testId = "console-list-item";

const defaultProps: ConsoleListItemProps = {
  title: "any_title",
  subtitle: "any_subtitle",
  iconName: "edit",
};

const makeSut = (props?: Partial<ConsoleListItemProps>) =>
  render(<ConsoleListItem {...defaultProps} {...props} />);

describe("<ConsoleListItem>", () => {
  it("should render component", () => {
    const wrapper = makeSut();

    expect(wrapper).toMatchSnapshot();
  });

  describe("when it has a modifier", () => {
    const baseClasses = "console-list-item console-list-item--";

    it("should render with the is-active modifier", () => {
      const { getByTestId } = makeSut({ isActive: true });

      expect(getByTestId(testId)).toHaveClass(`${baseClasses}is-active`);
    });

    it("should render with the is-disabled modifier", () => {
      const { getByTestId } = makeSut({ isDisabled: true });

      expect(getByTestId(testId)).toHaveClass(`${baseClasses}is-disabled`);
    });

    it("should render with the is-connector-disabled modifier", () => {
      const { getByTestId } = makeSut({ isConnectorDisabled: true });

      expect(getByTestId(testId)).toHaveClass(`${baseClasses}is-connector-disabled`);
    });

    it("should render with the is-answered modifier", () => {
      const { getByTestId } = makeSut({ isAnswered: true });

      expect(getByTestId(testId)).toHaveClass(`${baseClasses}is-answered`);
    });
  });

  describe("when has a link or custom component", () => {
    it("should render as an anchor element", () => {
      const { getByTestId } = makeSut({
        href: "https://www.mesalva.com",
      });

      const component = getByTestId(testId);

      expect(component).toBeInstanceOf(HTMLAnchorElement);
      expect(component).toHaveAttribute("href", "https://www.mesalva.com");
    });

    it("should render as a custom component", () => {
      const LinkComponent = "button";

      const { getByTestId } = makeSut({
        LinkComponent,
      });

      const component = getByTestId(testId);

      expect(component.tagName).toBe("BUTTON");
    });
  });
});
