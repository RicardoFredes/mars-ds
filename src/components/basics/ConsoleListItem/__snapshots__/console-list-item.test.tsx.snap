// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConsoleListItem> should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="console-list-item"
        data-testid="console-list-item"
      >
        <div
          class="console-list-item__icon-container"
        >
          <span
            class="icon"
            data-testid="edit"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M16.66 4.59998L19.4 7.33998C20.166 8.06975 20.1973 9.28164 19.47 10.05L10.47 19.05C10.144 19.3732 9.71679 19.5745 9.26 19.62L5.09 20H5C4.73419 20.0015 4.47872 19.8972 4.29 19.71C4.07903 19.4998 3.97272 19.2065 4 18.91L4.43 14.74C4.4755 14.2832 4.67676 13.8559 5 13.53L14 4.52998C14.7758 3.87453 15.9197 3.90464 16.66 4.59998ZM13.32 7.99998L16 10.68L18 8.72998L15.27 5.99998L13.32 7.99998Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
        <div
          class="console-list-item__content"
        >
          <h2
            class="text text--size-md"
          >
            any_title
          </h2>
          <span
            class="caption console-list-item__subtitle"
          >
            any_subtitle
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="console-list-item"
      data-testid="console-list-item"
    >
      <div
        class="console-list-item__icon-container"
      >
        <span
          class="icon"
          data-testid="edit"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M16.66 4.59998L19.4 7.33998C20.166 8.06975 20.1973 9.28164 19.47 10.05L10.47 19.05C10.144 19.3732 9.71679 19.5745 9.26 19.62L5.09 20H5C4.73419 20.0015 4.47872 19.8972 4.29 19.71C4.07903 19.4998 3.97272 19.2065 4 18.91L4.43 14.74C4.4755 14.2832 4.67676 13.8559 5 13.53L14 4.52998C14.7758 3.87453 15.9197 3.90464 16.66 4.59998ZM13.32 7.99998L16 10.68L18 8.72998L15.27 5.99998L13.32 7.99998Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
      <div
        class="console-list-item__content"
      >
        <h2
          class="text text--size-md"
        >
          any_title
        </h2>
        <span
          class="caption console-list-item__subtitle"
        >
          any_subtitle
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
